pkg/imagesize/imagesize.go

```diff
@@ -0,0 +1,164 @@
+	}
+
+	scheme := runtime.NewScheme()
+	_ = v1.AddToScheme(scheme)
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(1)
+	eventBroadcaster.StartRecordingToSink(&typedv1.EventSinkImpl{Interface: clientset.CoreV1().Events("")})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme, v1.EventSource{
+		Component: "warm-metal",
+	})
+
+	Warner = &warner{
+		recorder:     eventRecorder,
+		broadcaster:  eventBroadcaster,
+		clientSet:    clientset,
+		maxImageSize: maxImageSize,
+	}
+}
+
+// Cleanup performs cleanup activity (like shutting down the event broadcaster)
+func (w *warner) Cleanup() {
+	w.broadcaster.Shutdown()
+}
+
+// Warn creates the warning event object for the pod
+func (w *warner) Warn(current *resource.Quantity, pod string, namespace string) error {
+	if current.Cmp(*w.maxImageSize) <= 0 {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout)
+	defer cancel()
+	// TODO(vadasambar): set a timeout for the context here
+	po, err := w.clientSet.CoreV1().Pods(namespace).Get(ctx, pod, metav1.GetOptions{})
+	if err != nil {
+		e := fmt.Errorf("failed to get pod info to log event: %v, pod: %s, namespace: %s", err, pod, namespace)
+		klog.Error(e)
+		return e
+	}
+
+	c := fmt.Sprintf("%.2f MiB\n", float64(current.Value()/(1024.0*1024.0)))
+
+	w.recorder.Event(po, v1.EventTypeWarning,
+		"ImageSizeTooLarge", fmt.Sprintf("Image size '%v' is too large (max image size: %v)", strings.TrimSpace(c), w.maxImageSize.String()))
+
+	return nil
+}
+
+// GetImageSize gets the image size from the container registry
+func (w *warner) GetImageSize(creds *cri.AuthConfig, image docker.Named) (*resource.Quantity, error) {
+	resolver := func(ctx context.Context, index *registrytypes.IndexInfo) registrytypes.AuthConfig {
+		if creds == nil {
+			return registrytypes.AuthConfig{}
+		}
+		return registrytypes.AuthConfig{
+			Username:      creds.Username,
+			Password:      creds.Password,
+			Auth:          creds.Auth,
+			ServerAddress: creds.ServerAddress,
+			IdentityToken: creds.IdentityToken,
+			RegistryToken: creds.RegistryToken,
+		}
+	}
+	c := client.NewRegistryClient(resolver, "warm-metal", false)
+
+	return w.fetchImageSize(c, image)
+}
+
+// fetchImageSize gets the image size from the container registry
+func (w *warner) fetchImageSize(c client.RegistryClient, image reference.Named) (*resource.Quantity, error) {
+	var mlist []manifesttypes.ImageManifest
+	imageManifest, err := c.GetManifest(context.Background(), image)
+	if err == nil {
+		mlist = []manifesttypes.ImageManifest{
+			imageManifest,
+		}
+
+	} else {
+		mlist, err = c.GetManifestList(context.Background(), image)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get image manifest and manifest list: %v, image: %s", err, image)
+		}
+	}
+
+	var parsedSize *resource.Quantity
+
+	for _, m := range mlist {
+		// TODO(vadasambar): support other OS platforms
```

Comment:
Do we want to fix this in the same PR or create a separate issue for tracking this?